crates/ra_proc_macro/src/process.rs

```diff
@@ -0,0 +1,209 @@
+//! Handle process life-time and message passing for proc-macro client
+
+use crossbeam_channel::{bounded, Receiver, Sender};
+use ra_tt::Subtree;
+
+use crate::msg::{ErrorCode, Message, Request, Response, ResponseError};
+use crate::rpc::{ExpansionResult, ExpansionTask, ListMacrosResult, ListMacrosTask, ProcMacroKind};
+
+use io::{BufRead, BufReader};
+use std::{
+    convert::{TryFrom, TryInto},
+    io::{self, Write},
+    path::{Path, PathBuf},
+    process::{Child, Command, Stdio},
+};
+
+#[derive(Debug, Default)]
+pub(crate) struct ProcMacroProcessSrv {
+    inner: Option<Sender<Task>>,
+}
+
+#[derive(Debug)]
+pub(crate) struct ProcMacroProcessThread {
+    // XXX: drop order is significant
+    sender: SenderGuard,
+    handle: jod_thread::JoinHandle<()>,
+}
+
+#[derive(Debug)]
+struct SenderGuard(pub Sender<Task>);
+
+impl std::ops::Drop for SenderGuard {
+    fn drop(&mut self) {
+        let _ = self.0.send(Task::Close);
```

Comment:
When working rust rust channels, it's better not to send `close` events explicitelly, and instead just rely on the channel returning `None` from `.recv` once it is closed.

Reply:
Fixed : I changed to  `ProcMacroProcessSrv` holding a weak pointer of the sender. such that the only sender is holded by `ProcMacroProcessThread`.